src/mds/OpenFileTable.cc

```diff
@@ -0,0 +1,886 @@
+  uint64_t log_seq;
+  MDSInternalContextBase *fin;
+  MDSRank *get_mds() override { return oft->mds; }
+public:
+  C_IO_OFT_Save(OpenFileTable *t, uint64_t s, MDSInternalContextBase *c) :
+    oft(t), log_seq(s), fin(c) {}
+  void finish(int r) {
+    oft->_commit_finish(r, log_seq, fin);
+  }
+};
+
+void OpenFileTable::_commit_finish(int r, uint64_t log_seq, MDSInternalContextBase *fin)
+{
+  dout(10) << __func__ << " log_seq " << log_seq << dendl;
+  if (r < 0) {
+    mds->handle_write_error(r);
+    return;
+  }
+
+  assert(log_seq <= committing_log_seq);
+  assert(log_seq >= committed_log_seq);
+  committed_log_seq = log_seq;
+  num_pending_commit--;
+
+  if (fin)
+    fin->complete(r);
+}
+
+void OpenFileTable::commit(MDSInternalContextBase *c, uint64_t log_seq, int op_prio)
+{
+  dout(10) << __func__ << " log_seq " << log_seq << dendl;
+  assert(log_seq >= committing_log_seq);
+  committing_log_seq = log_seq;
+
+  omap_version++;
+
+  C_GatherBuilder gather(g_ceph_context,
+			 new C_OnFinisher(new C_IO_OFT_Save(this, log_seq, c),
+					  mds->finisher));
+
+  SnapContext snapc;
+  object_t oid = get_object_name();
+  object_locator_t oloc(mds->mdsmap->get_metadata_pool());
+
+  const unsigned max_write_size = mds->mdcache->max_dir_commit_size;
+  unsigned write_size = 0;
+  unsigned journal_idx = 0;
+  std::set<string> journal_keys;
+  std::map<string, bufferlist> to_update, journaled_update;
+  std::set<string> to_remove, journaled_remove;
+
+  auto journal_func = [&]() {
+    ObjectOperation op;
+    op.priority = op_prio;
+
+    if (clear_on_commit) {
+      op.omap_clear();
+      op.set_last_op_flags(CEPH_OSD_OP_FLAG_FAILOK);
+      clear_on_commit = false;
+    }
+
+    if (journal_idx == 0) {
+      assert(journal_state == JOURNAL_NONE);
+      journal_state = JOURNAL_START;
+      bufferlist header;
+      _encode_header(header);
+      op.omap_set_header(header);
+    }
+
+    bufferlist bl;
+    encode(omap_version, bl);
+    encode(to_update, bl);
+    encode(to_remove, bl);
+
+    char key[32];
+    snprintf(key, sizeof(key), "_journal.%x", journal_idx++);
+    journal_keys.insert(key);
+    std::map<string, bufferlist> tmp_map;
+    tmp_map[key].swap(bl);
+    op.omap_set(tmp_map);
+
+    mds->objecter->mutate(oid, oloc, op, snapc, ceph::real_clock::now(), 0,
+			  gather.new_sub());
+
+    for (auto& it : to_update)
+      journaled_update[it.first].swap(it.second);
+    journaled_remove.insert(to_remove.begin(), to_remove.end());
+
+    to_update.clear();
+    to_remove.clear();
+  };
+
+  auto commit_func = [&](bool update_header) {
+    ObjectOperation op;
+    op.priority = op_prio;
+
+    if (clear_on_commit) {
+      op.omap_clear();
+      op.set_last_op_flags(CEPH_OSD_OP_FLAG_FAILOK);
+      clear_on_commit = false;
+    }
+
+    if (update_header) {
+      bufferlist header;
+      _encode_header(header);
+      op.omap_set_header(header);
+    }
+
+    if (!to_update.empty())
+      op.omap_set(to_update);
```

Comment:
Note: I'm going through this PR by commit (currently: 8e2972e8c0a1b01d62ec26c415a1d2fa26065068) and wanted to leave a comment here even though the final code is different.

Am I correct in understanding that the openfile table object (and there is only one right?) will have an omap key for every open file? I would think that's a problem as we should not have more than ~100k omap key/value pairs (same problem as dirfrags)?

Reply:
omap interface face already support partial read, I don't see any problem for large omap

Reply:
Well, there's a couple possible issues:
1) too-large objects are just a problem, whether it's omap or not.
2) the MDS has to read in a directory at a time, and then immediately expire stuff if that pushes it over its cache limits.

So, (2) is not an issue here. (1) might be, depending on how large it's possible for this entry to get. I don't know the typical numbers at which it becomes an issue.

Does ordering matter for these? If not, a basic chunking strategy where you create a new object every N open files seems like it would be pretty simple. (NB: I have not looked at any of the code in this PR.)

Reply:
Chunking the omap over several objects is possible but would require extra bookkeeping to track in which object an Anchor is stored so updates can be applied.

Reply:
key data is 35 bytes for now, (key name  + key data) is about 50 bytes. The omap is about 50M when there are 1 million open files. It's not that large